be/src/exec/stream/scan/stream_scan_operator.h

```diff
@@ -16,31 +16,126 @@
+        _is_epoch_finished = true;
+        return Status::OK();
+    }
+
+    bool is_epoch_finished() const override {
+        if (_is_epoch_finished) {
+            return true;
+        }
+        // if storage layer returns an error, we should make sure `pull_chunk` has a chance to get it
+        if (!_get_scan_status().ok()) {
+            return false;
+        }
+
+        // Any io task is running or needs to run.
+        if (_num_running_io_tasks > 0 || !_morsel_queue->empty() || !_old_chunk_sources.empty()) {
+            return false;
+        }
+
+        // Any shared chunk source from other ScanOperator
+        if (has_shared_chunk_source()) {
+            return false;
+        }
+
+        // Remain some data in the buffer
+        if (num_buffered_chunks() > 0) {
+            return false;
+        }
+
+        // This scan operator is epoch finished, if no more io tasks are running
+        // or need to run, and all the read chunks are consumed.
+        return true;
+    }
+
+    Status reset_epoch(RuntimeState* runtime_state) override {
+        _is_epoch_finished = false;
+        _is_epoch_start = true;
+        _current_epoch_info = _stream_epoch_manager->epoch_info();
+
+        // reset chunkSource status
+        // swap chunk queue
+        int chunk_source_size = _closed_chunk_sources.size();
+        for (int i = 0; i < chunk_source_size; i++) {
+            ChunkSourcePtr chunk_source_ptr = _closed_chunk_sources.front();
+            _closed_chunk_sources.pop();
+            chunk_source_ptr->reset_status();
+            _old_chunk_sources.push(chunk_source_ptr);
+
+            auto tablet_id = chunk_source_ptr->get_lane_owner();
+            auto binlog_offset =
+                    _stream_epoch_manager->get_binlog_offset(runtime_state->fragment_instance_id(), _id, tablet_id);
+            chunk_source_ptr->set_stream_offset(binlog_offset->tablet_version, binlog_offset->lsn);
+            chunk_source_ptr->set_epoch_limit(_current_epoch_info.max_scan_rows, _current_epoch_info.max_exec_millis);
+        }
+
+        return Status::OK();
+    }
+
+    bool is_finished() const override { return _stream_epoch_manager && _stream_epoch_manager->is_finished(); }
+
+    Status set_epoch_finished(RuntimeState* state) override {
+        // todo report every chunk source with real offset
+        int chunk_source_size = _closed_chunk_sources.size();
+        for (int i = 0; i < chunk_source_size; i++) {
+            BinlogOffset binlog_offset{0, 2, 2};
+            _stream_epoch_manager->update_binlog_offset(state->fragment_instance_id(), _id, 0, binlog_offset);
+        }
+        return Status::OK();
+    }
+
+    Status do_prepare(RuntimeState* state) override {
+        ConnectorScanOperator::do_prepare(state);
+        _stream_epoch_manager = state->query_ctx()->stream_epoch_manager();
+        DCHECK(_stream_epoch_manager);
+        return Status::OK();
+    }
+
+private:
+    BinlogOffset* _offset;
```

Comment:
why not maintain the binlog offset in BinlogDataSource, but in the Operator? Because they are not 1:1 mapping